### ISSUE TITLE
A suggestion for AFHMM and AFHMM SAC

### ISSUE THREAD

**Issue by klemenjak** [nilmtk/nilmtk-contrib#27]
Hi there,

this is not an issue but rather something I was wondering about. 

While running AFHMM and AFHMM + SAC in my experiments, I recognized that those approaches really occupy a lot of CPU power. Even our simulation servers struggle with those algorithms sometimes.

Have you thought about providing a single-thread implementation of **disaggregate_chunk** ? Is it possible to limit the number of created threads in an easy manner?

Feel free to ignore this :)

Best,
C

**Comment by Rithwikksvr** [comment]
Hey @klemenjak 

If you increase the `'time_period'` parameter value of the AFHMM or AFHMM-SAC, you can decrease the number of threads.

First, we break the input sequence into multiple chunks of length `'time_period'`. Each of those chunks is given a thread.

So, by increase the value of `'time_period'`, you can decrease the threads.

It takes more time if you increase the `'time_period'` value. 

For our buildsys-experiments, we used a powerful CPU system with many threads. We were using `'time_period':720`, and even then it took a lot of time. 

Hope this was helpful!

**Comment by klemenjak** [comment]
Hi,

thank you so much! I will give it a shot!

Cheers,
C

**Comment by Wusir2018** [comment]
Hi @klemenjak 
Could you give me a example with `AFHMM` and `AFHMM_SAC`?  I can't run them correctly.
Best,
Wu

**Comment by klemenjak** [comment]
Hi @Wusir2018,

I haven't tested that so far. Sorry.

**Comment by Wusir2018** [comment]
Hey @klemenjak 
It doesn't matter. Thank you!

**Comment by Rithwikksvr** [comment]
HI @Wusir2018 

What exactly you aren't able to run?

**Comment by Wusir2018** [comment]
**Hey** @Rithwikksvr ,
**Thank you for your reply !**
**I can run `mean`, but I can't run `AFHMM` and `AFHMM SAC`. This is my code:**
![1](https://user-images.githubusercontent.com/51500032/83141851-df90f300-a122-11ea-8848-3a9cc7ba51f7.png)
![2](https://user-images.githubusercontent.com/51500032/83142163-f46d8680-a122-11ea-8f03-8b2ed272a440.png)

**But it stopped here:**
![3](https://user-images.githubusercontent.com/51500032/83142544-35fe3180-a123-11ea-839b-cbb7f4d0a2f5.png)

**Comment by Rithwikksvr** [comment]
Hey @Wusir2018 

The  AFHMM and AFHMM-SAC take significantly longer times for generating the  testing predictions. 

It takes a couple of hours to compute predictions for a single day

Also, us use a system with more CPUs

**Comment by Wusir2018** [comment]
Hey @Rithwikksvr ,
Thank you for your reply. It helps me a lot. It's my fault and my computer is poor.
I have a problem about `RNN,WindowGRU` with `AMPDs data`.
I can run `REDD data` with `RNN,WindowGRU`.
But I can't run `AMPDs data`, I think the sampling frequency of `AMPDs data` causes this problem. How can I solve this problem?

This is my code:
![1](https://user-images.githubusercontent.com/51500032/83146121-672d3080-a128-11ea-9318-acf977ae4266.png)
![2](https://user-images.githubusercontent.com/51500032/83146146-71e7c580-a128-11ea-98fd-eb1392e60902.png)

This is my error
![3](https://user-images.githubusercontent.com/51500032/83146197-85932c00-a128-11ea-84bd-567fbc28fd88.png)
![4](https://user-images.githubusercontent.com/51500032/83146219-8fb52a80-a128-11ea-844f-cd1587b1a4ce.png)

**Comment by Rithwikksvr** [comment]
I'll run it once and get back to you by tomorrow

**Comment by Wusir2018** [comment]
Hi @Rithwikksvr ,
Thank you very much !
Best wishes !

**Comment by chaosimple** [comment]
Hi @Rithwikksvr,

I am new to this area, so I don't quite clear about what exactly does the AFHMM algorithm refer to?
Is it refer to the "Additive FHMM" or the "Approximate FHMM"?

**Comment by MingjunZhong** [comment]
@chaosimple I think it was referring to the additive FHMM model and the method solving it was an approximate method employing Second-Order Cone Programming (SOCP).

**Comment by PMeira** [comment]
@chaosimple Check the README on the front page of the repository. The paper linked there should list the references: https://github.com/nilmtk/nilmtk-contrib/blob/master/README.md

(right now, the code is really missing some basic info as docstrings or comments)

**Comment by levaphenyl** [comment]
Note: processes of AFHMM and AFHMM+SAC should be [daemonic](https://docs.python.org/3/library/multiprocessing.html#multiprocessing.Process.daemon). In case the OS kills the main process, e.g. because of memory pressure, the current implementation leaves tons of wild orphaned processes behind.

**Comment by Rithwikksvr** [comment]
But @levaphenyl , do you know any fix we can use for thiis?

**Comment by levaphenyl** [comment]
> But @levaphenyl , do you know any fix we can use for this?

Yes, @Rithwikksvr. I use `multiprocessing.Pool` in my [PR](https://github.com/nilmtk/nilmtk-contrib/pull/37).

@klemenjak, I refactored AFHMM+SAC to use all CPUs - 1. Is it compatible with your original intent?
